core/utils/idgenerator.js

```diff
@@ -5,34 +5,74 @@
  */
 
 /**
- * @fileoverview Generator for unique element IDs.
- * For UUIDs use Blockly.utils.genUid. The ID generator should primarily be
- * used for IDs that end up in the DOM.
+ * @fileoverview Generators for unique IDs.
  * @author samelh@google.com (Sam El-Husseini)
  */
 'use strict';
 
 /**
- * @name Blockly.utils.IdGenerator
+ * @name Blockly.utils.idGenerator
  * @namespace
  */
-goog.module('Blockly.utils.IdGenerator');
+goog.module('Blockly.utils.idGenerator');
 goog.module.declareLegacyNamespace();
 
 
+/** 
+ * Namespace object for internal implementations we want to be able to
+ * stub in tests.
+ */
+const internal = {};
+exports.TEST_ONLY = internal;
+
 /**
  * Next unique ID to use.
  * @type {number}
  */
 let nextId = 0;
 
 /**
- * Gets the next unique ID.
+ * Generate the next unique element IDs.
  * IDs are compatible with the HTML4 id attribute restrictions:
  * Use only ASCII letters, digits, '_', '-' and '.'
+ *
+ * For UUIDs use genUid (below) instead; this ID generator should
+ * primarily be used for IDs that end up in the DOM.
+ *
  * @return {string} The next unique identifier.
  */
 const getNextUniqueId = function() {
   return 'blockly-' + (nextId++).toString(36);
 };
 exports.getNextUniqueId = getNextUniqueId;
+
+/**
+ * Legal characters for the universally unique IDs.  Should be all on
+ * a US keyboard.  No characters that conflict with XML or JSON.
+ * Requests to remove additional 'problematic' characters from this
+ * soup will be denied.  That's your failure to properly escape in
+ * your own environment.  Issues #251, #625, #682, #1304.
+ */
+const soup = '!#$%()*+,-./:;=?@[]^_`{|}~' +
+    'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
+
+ /** @return {string} */
+internal.genUid = function() {
+  const length = 20;
+  const soupLength = soup.length;
+  const id = [];
+  for (let i = 0; i < length; i++) {
+    id[i] = soup.charAt(Math.random() * soupLength);
+  }
+  return id.join('');
+};
+
+/**
+ * Generate a random unique ID.  This should be globally unique.
```

Comment:
Move this documentation to `internal.genUid` and then either duplicate it or use `@see`. Better to have it attached to the full implementation.

Reply:
OK.  I'd put it here so that someone browsing auto-generated API docs for the module would see it attached to the part of the API we want them to use, but I've re-arranged it as you suggested.